File: static/js/blurhash.js
// Декодер BlurHash — парный к internal/blurhash.go (стандартный алгоритм
// blurha.sh): base83, DC 4 символа, AC по 4 символа. Отдаёт ImageData
// для отрисовки плейсхолдера на canvas до загрузки полной картинки.

const B83 = '0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz#$%*+,-.:;=?@[]^_{|}~';

function b83Decode(str, length) {
  let value = 0;
  for (let i = 1; i <= length; i++) {
    const c = B83.indexOf(str[i - 1]);
    if (c < 0) return -1;
    value += c * Math.pow(83, length - i);
  }
  return value;
}

function srgbToLinear(v) {
  const x = v / 255;
  return x <= 0.04045 ? x / 12.92 : Math.pow((x + 0.055) / 1.055, 2.4);
}

function linearToSrgb(v) {
  const x = Math.max(0, Math.min(1, v));
  return x <= 0.0031308 ? Math.round(x * 12.92 * 255 + 0.5) : Math.round((1.055 * Math.pow(x, 1 / 2.4) - 0.055) * 255 + 0.5);
}

function signPow(v, exp) {
  return (v < 0 ? -1 : 1) * Math.pow(Math.abs(v), exp);
}

export function decodeBlurHash(hash, width, height) {
  if (!hash || hash.length < 6) return null;
  const sizeFlag = b83Decode(hash[0], 1);
  if (sizeFlag < 0) return null;
  const numY = Math.floor(sizeFlag / 9) + 1;
  const numX = (sizeFlag % 9) + 1;
  const quantMax = b83Decode(hash[1], 1);
  if (quantMax < 0) return null;
  const actualMax = (quantMax + 1) / 166;
  if (hash.length !== 2 + 4 + 4 * (numX * numY - 1)) return null;

  const colours = [];
  for (let i = 0; i < numX * numY; i++) {
    if (i === 0) {
      const v = b83Decode(hash.substring(2, 6), 4);
      if (v < 0) return null;
      colours.push([srgbToLinear(v >> 16), srgbToLinear((v >> 8) & 255), srgbToLinear(v & 255)]);
    } else {
      const v = b83Decode(hash.substring(4 + i * 4, 8 + i * 4), 4);
      if (v < 0) return null;
      const qR = Math.floor(v / 361), qG = Math.floor(v / 19) % 19, qB = v % 19;
      colours.push([
        signPow((qR - 9) / 9, 0.5) * actualMax,
        signPow((qG - 9) / 9, 0.5) * actualMax,
        signPow((qB - 9) / 9, 0.5) * actualMax,
      ]);
    }
  }

  const cosX = [];
  for (let i = 0; i < numX; i++) cosX[i] = new Float32Array(width);
  const cosY = [];
  for (let j = 0; j < numY; j++) cosY[j] = new Float32Array(height);
  for (let i = 0; i < numX; i++) for (let x = 0; x < width; x++) cosX[i][x] = Math.cos(Math.PI * i * x / width);
  for (let j = 0; j < numY; j++) for (let y = 0; y < height; y++) cosY[j][y] = Math.cos(Math.PI * j * y / height);

  const pixels = new Uint8ClampedArray(width * height * 4);
  for (let y = 0; y < height; y++) {
    for (let x = 0; x < width; x++) {
      let r = 0, g = 0, b = 0;
      for (let j = 0; j < numY; j++) {
        for (let i = 0; i < numX; i++) {
          const basis = (i === 0 && j === 0 ? 1 : 2) * cosX[i][x] * cosY[j][y];
          const c = colours[j * numX + i];
          r += basis * c[0];
          g += basis * c[1];
          b += basis * c[2];
        }
      }
      const o = 4 * (y * width + x);
      pixels[o] = linearToSrgb(r);
      pixels[o + 1] = linearToSrgb(g);
      pixels[o + 2] = linearToSrgb(b);
      pixels[o + 3] = 255;
    }
  }
  return new ImageData(pixels, width, height);
}

File: internal/blurhash.go
package internal

import (
	"image"
	"math"

	"github.com/disintegration/imaging"
)

// ── BlurHash (кодирование) ───────────────────────────────────────────────
// Стандартный алгоритм (https://blurha.sh, компоненты 3×3): картинка в
// linear-RGB раскладывается косинус-базисом, коэффициенты пакуются в
// компактную base83-строку. Декодер — на клиенте (static/js/blurhash.js):
// строка < 30 байт вместо тысяч байт превью-картинки.

const base83Chars = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz#$%*+,-.:;=?@[]^_{|}~"

const (
	blurhashNX = 3
	blurhashNY = 3
)

func base83Encode(value, length int) string {
	buf := make([]byte, length)
	digit := 1
	for i := length - 1; i >= 0; i-- {
		buf[i] = base83Chars[(value/digit)%83]
		digit *= 83
	}
	return string(buf)
}

func sRGBToLinear(v int) float64 {
	c := float64(v) / 255
	if c <= 0.04045 {
		return c / 12.92
	}
	return math.Pow((c+0.055)/1.055, 2.4)
}

func signPow(v, exp float64) float64 {
	if v < 0 {
		return -math.Pow(-v, exp)
	}
	return math.Pow(v, exp)
}

func linearToSRGB(v float64) int {
	x := math.Max(0, math.Min(1, v))
	if x <= 0.0031308 {
		return int(x*12.92*255 + 0.5)
	}
	return int((1.055*math.Pow(x, 1/2.4)-0.055)*255 + 0.5)
}

func encodeDC(f [3]float64) string {
	r := linearToSRGB(f[0])
	g := linearToSRGB(f[1])
	b := linearToSRGB(f[2])
	return base83Encode((r<<16)+(g<<8)+b, 4)
}

func encodeAC(f [3]float64, actualMax float64) string {
	quant := func(v float64) int {
		return int(math.Max(0, math.Min(18, math.Floor(signPow(v/actualMax, 0.5)*9+9.5))))
	}
	qr, qg, qb := quant(f[0]), quant(f[1]), quant(f[2])
	return base83Encode(qr*19*19+qg*19+qb, 4)
}

// BlurHashFile кодирует blurhash из файла; "" — файл не прочитался.
func BlurHashFile(path string) string {
	img, err := imaging.Open(path)
	if err != nil {
		return ""
	}
	return BlurHashEncode(img)
}

// BlurHashEncode возвращает blurhash строки картинки; "" — не получилось.
func BlurHashEncode(img image.Image) string {
	if img == nil {
		return ""
	}
	small := imaging.Fit(img, 32, 32, imaging.Lanczos)
	b := small.Bounds()
	w, h := b.Dx(), b.Dy()
	if w < 2 || h < 2 {
		return ""
	}
	// Пиксели в linear-RGB.
	lin := make([][][3]float64, h)
	for y := 0; y < h; y++ {
		lin[y] = make([][3]float64, w)
		for x := 0; x < w; x++ {
			r, g, bl, _ := small.At(b.Min.X+x, b.Min.Y+y).RGBA()
			lin[y][x] = [3]float64{
				sRGBToLinear(int(r >> 8)),
				sRGBToLinear(int(g >> 8)),
				sRGBToLinear(int(bl >> 8)),
			}
		}
	}
	// Косинус-коэффициенты (нормировка как в референс-реализации).
	factors := make([][3]float64, blurhashNX*blurhashNY)
	for j := 0; j < blurhashNY; j++ {
		for i := 0; i < blurhashNX; i++ {
			norm := 1.0
			if i != 0 || j != 0 {
				norm = 2
			}
			var f [3]float64
			for y := 0; y < h; y++ {
				for x := 0; x < w; x++ {
					basis := norm *
						math.Cos(math.Pi*float64(i)*float64(x)/float64(w)) *
						math.Cos(math.Pi*float64(j)*float64(y)/float64(h))
					p := lin[y][x]
					f[0] += basis * p[0]
					f[1] += basis * p[1]
					f[2] += basis * p[2]
				}
			}
			scale := 1.0 / float64(w*h)
			factors[j*blurhashNX+i] = [3]float64{f[0] * scale, f[1] * scale, f[2] * scale}
		}
	}
	dc := factors[0]
	ac := factors[1:]
	maxAc := 0.0
	for _, f := range ac {
		for _, v := range f {
			if math.Abs(v) > maxAc {
				maxAc = math.Abs(v)
			}
		}
	}
	var quantMax int
	if maxAc > 0 {
		quantMax = int(math.Max(0, math.Min(82, math.Floor(maxAc*166-0.5))))
	}
	actualMax := float64(quantMax+1) / 166

	out := base83Encode((blurhashNX-1)+(blurhashNY-1)*9, 1)
	out += base83Encode(quantMax, 1)
	out += encodeDC(dc)
	for _, f := range ac {
		out += encodeAC(f, actualMax)
	}
	return out
}
